Add nightly turtlebot

clalancette opened the pull request:
The change I really wanted to do was to add a nightly build job for turtlebot.  However, I found the way the current jobs are created to be confusing, so I also rewrote the loop in the first patch.  See the commit message for the first patch for an explanation.

I tested this in a dry-run locally; the only things that were going to be different from before were a small typo fix in the nightly_*_repeated jobs (an extra space at the end of the spec), and the creation of the two new nightly turtlebot jobs.

dirk-thomas commented:
Please double check the current configuration by comparing the newly generated config with the existing jobs (without actually overwriting them). You will notice that the patch has a couple of bugs atm.

clalancette commented:
Ah, good call.  I missed that.

dhood commented on create_jenkins_job.py:
```diff
@@ -114,73 +116,87 @@ def main(argv=None):
+                'turtlebot_demo': True,
+                'supplemental_repos_url': 'https://raw.githubusercontent.com/ros2/turtlebot2_demo/master/turtlebot2_demo.repos',
+            })
+            create_job(os_name, 'nightly_turtlebot-demo-%s' % (os_name), 'ci_job.xml.em', {
```
this needs to set turtlebot_demo and supplemental_repos_url, right? I'd also recommend the name include debug for consistency

mikaelarguedas replied:
`'turtlebot_demo': True,` << I think that this is needed as well

clalancette replied:
Yep, you are exactly right, totally missed.  Thanks, I've added that now, along with the "debug" name.  Please take another look when you have a moment.

mikaelarguedas commented on create_jenkins_job.py:
```diff
@@ -114,73 +116,89 @@ def main(argv=None):
-        configure_job(jenkins, job_name, job_config, **jenkins_kwargs)
+        short_name = 'nightly_' + os_name + '_debug'
+        if os_name == 'windows':
+            short_name = 'nightly_win_deb'
```
Is it necessary to change the logic of how the names shorten for windows ? (same everywhere the job names are hardcoded)

clalancette replied:
Not really necessary, no, but the previous way seemed overly complicated to me.  It was, in effect, hard-coding the short names, just in a less obvious way.

mikaelarguedas replied:
the previous logic was creating the job_name the same way for everyone and then modifying it to make it shorter for windows. This new approach replace the variable content with a hard-coded string literal discarding the previous content so I think it's more likely that the naming becomes inconsistent if we change one but not the other

mikaelarguedas commented on create_jenkins_job.py:
```diff
@@ -114,73 +116,89 @@ def main(argv=None):
-            job_config = expand_template('ci_job.xml.em', job_data)
-            configure_job(jenkins, job_name, job_config, **jenkins_kwargs)
-            job_data['enable_c_coverage_default'] = 'false'
+            create_job(os_name, 'nightly_linux_coverage', 'ci_job.xml.em', {
```
Same remark asabove, I'm not sure hardcoding the job names is a good idea. If we want to create a coverage job for osx we will need to create another conditional bloc while we could just us this one if it stayed as it was before

clalancette replied:
OK, yeah.  On this one, I agree with you.  I'll change this back to building it up from the os_name.

mikaelarguedas commented on create_jenkins_job.py:
```diff
@@ -114,73 +116,89 @@ def main(argv=None):
+                'supplemental_repos_url': 'https://raw.githubusercontent.com/ros2/turtlebot2_demo/master/turtlebot2_demo.repos',
+            })
+            create_job(os_name, 'nightly_turtlebot-demo_%s_debug' % (os_name), 'ci_job.xml.em', {
+                'cmake_build_type': 'Debug',
```
Note: Cartographer doesn't build in `Debug` AFAIK so I would expect this job to fail every night

clalancette replied:
Ah, I had forgotten about that fact, good call.  Will change it to release instead.

mikaelarguedas commented on create_jenkins_job.py:
```diff
@@ -114,73 +116,89 @@ def main(argv=None):
+        # configure turtlebot jobs on Linux only for now
+        if os_name in ['linux', 'linux-aarch64']:
+            create_job(os_name, 'ci_turtlebot-demo_%s' % (os_name), 'ci_job.xml.em', {
+                'cmake_build_type': 'None',
```
We usually never have None nightlies, is there a reason to have one for the turtlebot job ?

clalancette replied:
This was a cut-n-paste from the old turtlebot job.  I'm not sure where that configuration originally came from.

mikaelarguedas commented on create_jenkins_job.py:
```diff
@@ -114,73 +116,89 @@ def main(argv=None):
+
+        # configure turtlebot jobs on Linux only for now
+        if os_name in ['linux', 'linux-aarch64']:
+            create_job(os_name, 'ci_turtlebot-demo_%s' % (os_name), 'ci_job.xml.em', {
```
every other job uses a `job_name` variable and uses `+` for concatenation, can you do the same here to stay consistent?

clalancette replied:
Yeah, that's fine, I'll fix that.  This was a cut-n-paste from below, but I agree with you, it should be consistent.

clalancette commented:
> If the goal is to confirm that the code still builds I'm not sure we need to create several build types because it will take a lot of buildfarm time.

Sorry, I missed this bit before.  As it stands, this patch will create 2 new jobs: nightly_turtlebot-demo_linux and nightly_turtlebot-demo_linux-aarch64.  Are you worried that adding those two nightly builds will tax the build system, or were you concerned about something else?

mikaelarguedas commented:
> Sorry, I missed this bit before. As it stands, this patch will create 2 new jobs: nightly_turtlebot-demo_linux and nightly_turtlebot-demo_linux-aarch64. Are you worried that adding those two nightly builds will tax the build system, or were you concerned about something else?

I was concerned about mutliple turtlebot jobs **per architecture** (e.g. before one Debug and one None). If the current patch "create 2 new jobs: nightly_turtlebot-demo_linux and nightly_turtlebot-demo_linux-aarch64" that's good :+1:

mikaelarguedas commented on create_jenkins_job.py:
```diff
@@ -114,73 +116,89 @@ def main(argv=None):
+                'turtlebot_demo': True,
+                'supplemental_repos_url': 'https://raw.githubusercontent.com/ros2/turtlebot2_demo/master/turtlebot2_demo.repos',
+            })
+            create_job(os_name, 'nightly_turtlebot-demo_' + os_name, 'ci_job.xml.em', {
```
> I'd also recommend the name include debug for consistency

This still applies, we should add `_release` to the job name here

clalancette commented:
All right, I added release back to the name, and I've changed it up so that we are building the Windows names again.  I still made some changes in how we built the names; in particular I made it so that they have a common set of rules.  Ideally we'd apply those rules to all jobs, but that would currently replace some of the windows jobs with other jobs (ci_windows -> ci_win, packaging_windows -> packaging_win, etc).  While that is technically more consistent, I don't think we need to do that right now, so I just special cased the ones that were special-cased before.  @mikaelarguedas please take another look when you get a chance.

mikaelarguedas commented:
> I made it so that they have a common set of rules. Ideally we'd apply those rules to all jobs,

:+1: agreed that a common rule makes more sense but appreciate that you apply them only to the jobs that needed it to keep the naming. Ideally the rule of cropping after the 15th character should be shared as well rather than duplicated for all the jobs needing cropping

clalancette commented:
Thanks for the review.  I'm going to merge this and then deploy it.
